### PR TITLE
fix(rewards): collision penalty should not apply when controller issues the `STOP` action.

### LINKED ISSUE #363: Add test coverage variable speed and crash penalty.

### PR DESCRIPTION
**Pull request by CleverManu**
Collision penalty should only be applied when env has to intervene.

## Changes

Describe the changes you made.

## Related issues
Closes #363 

Link to every issue from the issue tracker (if any) you addressed in this PR.

## Checklist

- [ ] Tests are included for relevant behavior changes.
- [ ] Documentation is added in the [flatland-book](https://github.com/flatland-association/flatland-book) repo for relevant behavior changes.
- [ ] If you made important user-facing changes, describe them under the `[Unreleased]` tag in `CHANGELOG.md`.
- [ ] New package dependencies are declared in the `pyproject.toml` file.
  Requirement files have been updated by running `tox -e requirements`.
- [ ] Code works with all supported Python versions (3.10, 3.11, 3.12 and 3.13). Checks run with all supported version and are
  required to run successfully.
- [ ] Code is formatted according to PEP 8 (an IDE like PyCharm can do this for you).
- [ ] Technical guidelines listed in `CONTRIBUTING.md` are followed.

### CODE REVIEW

File: tests/test_rewards.py [hunk 1]
```diff
@@ -676,3 +681,167 @@ def test_ecml2026_rewards_normalization():
...
+
+    # 4th step: agents at (3,3) east and (3,4) west want to swap -> motion check stops agent 1
+    _, rewards, _, _ = env.step(forward)
+    assert agent_1.state_machine.previous_state == TrainState.MOVING
+    assert agent_1.state == TrainState.STOPPED
+    assert rewards[1][DefaultPenalties.COLLISION.value] == -1 * 1 * COLLISION_FACTOR, \
+        "An env-forced stop (head-on conflict) must incur the collision penalty proportional to speed"
+    # the agent winning the conflict resolution keeps moving unpenalized
+    assert agent_0.state == TrainState.MOVING
+    assert rewards[0][DefaultPenalties.COLLISION.value] == 0
```
Comment (chenkins) [inline tests/test_rewards.py]: @CleverManu Another bug?

File: flatland/envs/rewards.py
Comment (chenkins) [inline flatland/envs/rewards.py:1]: @CleverManu add test coverage for this edge case as well:

> When a train is already stopped, say a malfunction just ended, and its plan says to wait a bit longer before moving, we have to send DO_NOTHING and keep it stopped and then start it at the last moment, instead of letting it inch forward and brake again, because restarting a stopped train is free but braking a moving one is -250. 


https://github.com/darshanmakwana412/ecml2026/tree/main

Comment (CleverManu) [reply flatland/envs/rewards.py:1]: Isn't that just the regular "stop = penalty" problem we are trying to solve here? (According to my understanding it is)

File: tests/test_rewards.py (hunk 1 above)
Comment (CleverManu) [reply tests/test_rewards.py]: No, just claude writing correct code with very misleading comments.
I changed it and added two more steps in this test.

File: flatland/envs/rewards.py
Comment (chenkins) [reply flatland/envs/rewards.py:1]: State transitions might have their peculiarities... Maybe worth a test to document the behaviour and keep it regression-safe.

Comment (CleverManu) [reply flatland/envs/rewards.py:1]: added

File: tests/test_rewards.py [hunk 2]
```diff
@@ -701,3 +707,305 @@ def test_ecml2026_rewards_normalization():
...
+    assert agent.state == TrainState.STOPPED
+
+
+def test_no_collision_penalty_on_voluntary_stop():
+    """STOP_MOVING issued by the policy, braking reaches zero, no env intervention -> no penalty."""
+    rewards = BaseDefaultRewards(collision_factor=COLLISION_FACTOR)
+    agent, distance_map = _moving_agent()
+
+    signals = StateTransitionSignals(stop_action_given=True, new_speed_zero=True, movement_allowed=True)
+    _stop_moving_agent(agent, signals)
```
Comment (chenkins) [inline tests/test_rewards.py:838]: Is this the right level the test is written at: do we not want to test that the combination of step with an action (without tweaking state transition signals in the test) gives the expected rewards? I.e. we would need to add a second agent at the position we want to go to instead of `movement_allowed=False`. Or covered by "e2e" below?

File: tests/test_rewards.py [hunk 3]
```diff
@@ -701,3 +707,305 @@ def test_ecml2026_rewards_normalization():
...
+        assert rewards[0][DefaultPenalties.COLLISION.value] == 0
+    _, rewards, _, _ = env.step({0: RailEnvActions.MOVE_FORWARD})
+    assert agent.state == TrainState.MOVING
+    assert rewards[0][DefaultPenalties.COLLISION.value] == 0
+
+    # strategy 2 (stop-and-go): brake the restarted train voluntarily -- also free since the fix
+    _, rewards, _, _ = env.step({0: RailEnvActions.STOP_MOVING})
+    assert agent.state == TrainState.STOPPED
+    assert rewards[0][DefaultPenalties.COLLISION.value] == 0, \
+        "Braking a restarted train must cost the same as holding it stopped: nothing"
```
Comment (chenkins) [inline tests/test_rewards.py]: would have exptected strategy2 applied out of malfunction as well

File: tests/test_rewards.py [hunk 4]
```diff
@@ -701,3 +707,305 @@ def test_ecml2026_rewards_normalization():
...
+    assert rewards[2][DefaultPenalties.COLLISION.value] == -1 * 1 * COLLISION_FACTOR
+    assert rewards[0][DefaultPenalties.COLLISION.value] == 0  # leader itself moves freely
+
+    # steps 1..4: leader vacates a cell each tick, platoon moves in step -> no penalty for anyone
+    for _ in range(4):
+        _, rewards, _, _ = env.step(forward)
+        for i in range(3):
+            assert rewards[i][DefaultPenalties.COLLISION.value] == 0
+
+    # step 5: one full period later, the leader again fails to vacate its cell -> followers blocked again
```
Comment (chenkins) [inline tests/test_rewards.py:1101]: Should we add a second period to show the pattern continues, or at least step 6 to show that they can all run again?

File: tests/test_rewards.py (hunk 3 above)
Comment (CleverManu) [reply tests/test_rewards.py]: fixed

File: tests/test_rewards.py (hunk 4 above)
Comment (CleverManu) [reply tests/test_rewards.py:1101]: omitted for simplicity's sake (make_simple_rail would need to be changed to not be as "simple" anymore)

**Review by chenkins — APPROVED** [review]
